neqo-transport/src/recv_stream.rs

```diff
@@ -200,41 +200,46 @@ impl RxStreamOrderer {
             false
         };
 
-        // Now handle possible overlap with next entries
-        let mut to_remove = SmallVec::<[_; 8]>::new();
         let mut to_add = new_data;
+        if let Some(entry) = self.data_ranges.last_entry() {
```

Comment:
It took me a while to see this, but your change is pretty small.

This probably needs pictures, like the ones above.

This chunk is for any ranges that might start after the one that is being proposed.

In most cases, there are no ranges after the block, so you want a quick shortcut for that case.  That is your addition:

```
//  PPPPPP
//             NNNNNNN
```

Otherwise, you then have two lots of collapsing to do:

Firstly, any ranges for which the end of the old range is inside the new range, that's no longer needed:

```
//                PPPP
//             NNNNNNN
```

That might happen multiple times AND the open iterator means that we can't mutate the collection until the loop concludes, so we have to save those items that need to get dropped.  (`BTreeMap::drain(range)` would be awesome for that, but we don't get to have nice things, so we have a `SmallVec` for that.)

Finally, the end of this might be inside the next range, so this one can be truncated before adding it:

```
//                  PPPPPPP
//             NNNNNnn
```